core/src/test/scala/org/apache/spark/sql/delta/DeltaTestUtils.scala

```diff
@@ -204,6 +205,37 @@ object DeltaTestUtils extends DeltaTestUtilsBase {
       case tableName => tableName -> None
     }
   }
+
+  /**
+   * Implements an ordering where `x < y` iff both reader and writer versions of
+   * `x` are strictly less than those of `y`.
+   *
+   * Can be used to conveniently check that this relationship holds in tests/assertions
+   * without having to write out the conjunction of the two subconditions every time.
+   */
+  case object StrictProtocolOrdering extends PartialOrdering[Protocol] {
```

Comment:
I think changes in this file are already covered by `Protocol.canUpgradeTo`, i.e., if `A` can be upgraded to `B` then `B <= A`.
Not needed by this PR, but it would be nice to try that out in these tests.